### PR TITLE
Implement analysis and fix for OrderBy(x => x) to Order()

### LINKED ISSUE #1492: RCS1077 for OrderBy(x => x)
RCS1077 should recommend replacing `.OrderBy(x => x)` with `.Order()`.

### PR DESCRIPTION
**Pull request by BenjaminBrienen**
Fixes #1492 

TODO:
- [x] Unit test fails because the unit test is run with a version of .NET Core where `System.Linq` does not yet implement `IEnumerable<T>.Order()`.

### CODE REVIEW

**Comment by BenjaminBrienen** [comment]
@josefpihrt I added a conditional compilation on the unit test. If you want to handle this a different way, just let me know! :)

File: src/Analyzers.CodeFixes/CSharp/CodeFixes/OptimizeLinqMethodCallCodeFixProvider.cs [hunk 1]
```diff
@@ -565,6 +575,16 @@ private static Task<Document> CallOrderByDescendingInsteadOfOrderByAndReverseAsy
         return document.ReplaceNodeAsync(invocationInfo.InvocationExpression, newInvocationExpression, cancellationToken);
     }
 
+    private static Task<Document> CallOrderInsteadOfOrderByIdentityAsync(
+        Document document,
+        in SimpleMemberInvocationExpressionInfo invocationInfo,
+        CancellationToken cancellationToken)
+    {
+        InvocationExpressionSyntax newInvocationExpression = ChangeInvokedMethodName(invocationInfo.InvocationExpression, "Order");
```
Comment (josefpihrt) [inline src/Analyzers.CodeFixes/CSharp/CodeFixes/OptimizeLinqMethodCallCodeFixProvider.cs:583]: It's necessary to remove an argument. Currently, the code is changed to `Order(x => x)` which is not compilable.

File: src/Analyzers/CSharp/Analysis/OptimizeLinqMethodCallAnalysis.cs [hunk 2]
```diff
@@ -690,6 +690,43 @@ public static void AnalyzeOrderByAndReverse(SyntaxNodeAnalysisContext context, i
...
+    public static void AnalyzeOrderByIdentity(SyntaxNodeAnalysisContext context, in SimpleMemberInvocationExpressionInfo invocationInfo)
+    {
+        InvocationExpressionSyntax invocationExpression = invocationInfo.InvocationExpression;
+
+        if (context.SemanticModel.GetSymbolInfo(invocationExpression).Symbol is IMethodSymbol methodSymbol)
+        {
+            INamedTypeSymbol containingType = methodSymbol.ContainingType;
+            IMethodSymbol orderMethod = containingType.GetMembers("Order")
+                .OfType<IMethodSymbol>()
+                .FirstOrDefault(m => m.Parameters.Length is 0);
```
Comment (josefpihrt) [inline src/Analyzers/CSharp/Analysis/OptimizeLinqMethodCallAnalysis.cs:703]: A number of parameters should be 1, it's an extension method. You can also use extension method `FindMember`:

```
IMethodSymbol orderMethod = containingType.FindMember<IMethodSymbol>("Order", m => m.Parameters.Length == 1)
```

File: src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs [hunk 3]
```diff
@@ -959,6 +959,46 @@ void M()
...
+        IEnumerable<object> x = null;
+
+        x = x.Order();
+    }
+}
+";
+        await VerifyDiagnosticAndFixAsync(input, fix);
+#else
+    await VerifyNoDiagnosticAsync(input);
+#endif
```
Comment (josefpihrt) [inline src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs:999]: I created a PR which update target framework to .NET 8: https://github.com/dotnet/roslynator/pull/1523.

After that PR is merged and this branch is updated the test should pass without the conditional compilation.

**Review by josefpihrt — COMMENTED** [review]
Please update changelog.

File: src/Analyzers/CSharp/Analysis/OptimizeLinqMethodCallAnalysis.cs (hunk 2 above)
Comment (BenjaminBrienen) [reply src/Analyzers/CSharp/Analysis/OptimizeLinqMethodCallAnalysis.cs:703]: > A number of parameters should be 1, it's an extension method. You can also use extension method `FindMember`:
> 
> ```
> IMethodSymbol orderMethod = containingType.FindMember<IMethodSymbol>("Order", m => m.Parameters.Length == 1)
> ```

I couldn't figure out how exactly to call FindMember from `invocationInfo`.

File: src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs (hunk 3 above)
Comment (josefpihrt) [reply src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs:999]: @BenjaminBrienen I've merged the PR so you can update this branch.

Comment (BenjaminBrienen) [reply src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs:999]: Nice. Those other mistakes would've been caught with the updated target framework. Should I remove the conditional compilation and wait to merge until #1523 is merged?

Comment (josefpihrt) [reply src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs]: I've updated the branch so now the tests should pass.

Comment (BenjaminBrienen) [reply src/Tests/Analyzers.Tests/RCS1077OptimizeLinqMethodCallTests.cs]: After I fix the ArgumentsList, yeah. Currently trying to figure out how to remove arguments. At the moment, I only see methods for adding arguments and ArgumentsList doesn't have a public constructor.

**Comment by BenjaminBrienen** [comment]
At the moment, I only see methods for adding arguments and ArgumentsList doesn't have a public constructor.

**Comment by josefpihrt** [comment]
You can use combination of `With...` methods and `SyntaxFactory` methods:

```cs
newInvocationExpression = newInvocationExpression.WithArgumentList(
    newInvocationExpression.ArgumentList.WithArguments(
        SyntaxFactory.SeparatedList<ArgumentSyntax>()));
```

**Comment by BenjaminBrienen** [comment]
> You can use combination of `With...` methods and `SyntaxFactory` methods:
> 
> ```cs
> newInvocationExpression = newInvocationExpression.WithArgumentList(
>     newInvocationExpression.ArgumentList.WithArguments(
>         SyntaxFactory.SeparatedList<ArgumentSyntax>()));
> ```

Ah, I just found another way using RemoveNodes, but this works too! (and it is better)

**Comment by BenjaminBrienen** [comment]
I've doubled checked that all tests pass, so it should be ready for final review.

**Comment by josefpihrt** [comment]
Build is failing.

**Comment by BenjaminBrienen** [comment]
> Build is failing.

Oh wow, it fails even with a little info message. Good to know there are tight checks! Fixed.

**Review by josefpihrt — APPROVED** [review]

**Comment by josefpihrt** [comment]
Great PR! Thanks for the contribution! 👍

**Pull request merged**